// @bsv/wallet-toolbox/out/src/storage/__test/findStaleMerkleRootsKnex.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const TestUtilsWalletStorage_1 = require("../../../test/utils/TestUtilsWalletStorage");
const index_client_1 = require("../../index.client");
const StorageKnex_1 = require("../StorageKnex");
describe('findStaleMerkleRootsKnex test', () => {
    jest.setTimeout(99999999);
    const storages = [];
    const chain = 'test';
    const env = TestUtilsWalletStorage_1._tu.getEnv(chain);
    beforeAll(async () => {
        const localSQLiteFile = await TestUtilsWalletStorage_1._tu.newTmpFile('findstalemerkleroots.sqlite', false, false, false);
        const knexSQLite = TestUtilsWalletStorage_1._tu.createLocalSQLite(localSQLiteFile);
        storages.push(new StorageKnex_1.StorageKnex({
            ...StorageKnex_1.StorageKnex.defaultOptions(),
            chain,
            knex: knexSQLite
        }));
        if (env.runMySQL) {
            const knexMySQL = TestUtilsWalletStorage_1._tu.createLocalMySQL('findstalemerkleroots');
            storages.push(new StorageKnex_1.StorageKnex({
                ...StorageKnex_1.StorageKnex.defaultOptions(),
                chain,
                knex: knexMySQL
            }));
        }
        for (const storage of storages) {
            await storage.dropAllData();
            await storage.migrate('findStaleMerkleRoots tests', '1'.repeat(64));
            await storage.makeAvailable();
        }
    });
    afterAll(async () => {
        for (const storage of storages) {
            await storage.destroy();
        }
    });
    test('0 returns distinct stale merkle roots for the target height', async () => {
        for (const storage of storages) {
            await storage.dropAllData();
            await storage.migrate('findStaleMerkleRoots test reset', '1'.repeat(64));
            await storage.makeAvailable();
            const height = 777;
            const currentMerkleRoot = 'aa'.repeat(32);
            await storage.insertProvenTx(makeProvenTx(1, height, currentMerkleRoot));
            await storage.insertProvenTx(makeProvenTx(2, height, 'bb'.repeat(32)));
            await storage.insertProvenTx(makeProvenTx(3, height, 'bb'.repeat(32)));
            await storage.insertProvenTx(makeProvenTx(4, height, 'cc'.repeat(32)));
            await storage.insertProvenTx(makeProvenTx(5, height + 1, 'dd'.repeat(32)));
            const roots = await storage.findStaleMerkleRoots({
                height,
                merkleRoot: currentMerkleRoot
            });
            expect(roots.sort()).toEqual(['bb'.repeat(32), 'cc'.repeat(32)]);
        }
    });
});
function makeProvenTx(provenTxId, height, merkleRoot) {
    const now = new Date();
    return {
        created_at: now,
        updated_at: now,
        provenTxId: 0,
        txid: (0, index_client_1.randomBytesHex)(32),
        height,
        index: provenTxId,
        merklePath: [1, 2, 3, 4],
        rawTx: [5, 6, 7],
        blockHash: (0, index_client_1.randomBytesHex)(32),
        merkleRoot
    };
}
