// @bsv/wallet-toolbox/out/src/wab-client/auth-method-interactors/AuthMethodInteractor.js
"use strict";
/**
 * AuthMethodInteractor
 *
 * A base interface/class for client-side logic to interact with a server
 * for a specific Auth Method's flow (start, complete).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthMethodInteractor = void 0;
/**
 * Abstract client-side interactor for an Auth Method.
 *
 * Subclasses only need to set `methodType`; the HTTP calls to
 * `/auth/start` and `/auth/complete` are handled here.
 */
class AuthMethodInteractor {
    /**
     * Shared POST helper for auth endpoints.
     */
    async postAuth(serverUrl, endpoint, presentationKey, payload) {
        const res = await fetch(`${serverUrl}/auth/${endpoint}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                methodType: this.methodType,
                presentationKey,
                payload
            })
        });
        if (!res.ok) {
            return { success: false, message: `HTTP error ${res.status}` };
        }
        return res.json();
    }
    /**
     * Start the flow (e.g. request an OTP or create a session).
     */
    async startAuth(serverUrl, presentationKey, payload) {
        return this.postAuth(serverUrl, 'start', presentationKey, payload);
    }
    /**
     * Complete the flow (e.g. confirm OTP).
     */
    async completeAuth(serverUrl, presentationKey, payload) {
        return this.postAuth(serverUrl, 'complete', presentationKey, payload);
    }
}
exports.AuthMethodInteractor = AuthMethodInteractor;
