// @bsv/wallet-toolbox/out/src/monitor/tasks/TaskReviewUtxos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskReviewUtxos = void 0;
const sdk_1 = require("../../sdk");
const WalletMonitorTask_1 = require("./WalletMonitorTask");
/**
 * Use the reviewByIdentityKey method to review the utxos of a specific user by their identityKey.
 *
 * The task itself is disabled and will not run on a schedule; review must be triggered manually by calling reviewByIdentityKey.
 */
class TaskReviewUtxos extends WalletMonitorTask_1.WalletMonitorTask {
    constructor(monitor, triggerMsecs = 0, userLimit = 10, userOffset = 0, tags = ['release', 'all']) {
        super(monitor, TaskReviewUtxos.taskName);
        this.triggerMsecs = triggerMsecs;
        this.userLimit = userLimit;
        this.userOffset = userOffset;
        this.tags = tags;
    }
    trigger(_nowMsecsSinceEpoch) {
        return {
            run: false
        };
    }
    async runTask() {
        TaskReviewUtxos.checkNow = false;
        return 'TaskReviewUtxos is disabled; use reviewByIdentityKey instead.\n';
    }
    async reviewByIdentityKey(identityKey, mode = 'all') {
        const tags = ['release', ...(mode === 'all' ? ['all'] : [])];
        const vargs = {
            basket: sdk_1.specOpInvalidChange,
            tags,
            tagQueryMode: 'all',
            includeLockingScripts: false,
            includeTransactions: false,
            includeCustomInstructions: false,
            includeTags: false,
            includeLabels: false,
            limit: 0,
            offset: 0,
            seekPermission: false,
            knownTxids: []
        };
        return await this.storage.runAsStorageProvider(async (sp) => {
            const user = (await sp.findUsers({ partial: { identityKey } }))[0];
            if (!user) {
                return `identityKey ${identityKey} was not found\n`;
            }
            const auth = { userId: user.userId, identityKey: user.identityKey };
            const result = await sp.listOutputs(auth, vargs);
            if (result.totalOutputs === 0) {
                return `userId ${user.userId}: no invalid utxos found, ${user.identityKey}\n`;
            }
            const total = result.outputs.reduce((sum, output) => sum + output.satoshis, 0);
            return this.toUserLog(user, result.outputs, result.totalOutputs, total, tags);
        });
    }
    toUserLog(user, outputs, totalOutputs, total, tags) {
        const action = tags.includes('release') ? 'updated to unspendable' : 'found';
        const target = tags.includes('all') ? 'spendable utxos' : 'spendable change utxos';
        let log = `userId ${user.userId}: ${totalOutputs} ${target} ${action}, total ${total}, ${user.identityKey}\n`;
        for (const output of outputs) {
            log += `  ${output.outpoint} ${output.satoshis} now ${output.spendable ? 'spendable' : 'spent'}\n`;
        }
        return log;
    }
}
exports.TaskReviewUtxos = TaskReviewUtxos;
TaskReviewUtxos.taskName = 'ReviewUtxos';
TaskReviewUtxos.checkNow = false;
