// @bsv/wallet-toolbox/out/src/monitor/tasks/TaskReviewDoubleSpends.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskReviewDoubleSpends = void 0;
const Monitor_1 = require("../Monitor");
const WalletMonitorTask_1 = require("./WalletMonitorTask");
/**
 * Review recent reqs in terminal 'doubleSpend' state and move any false positives
 * back to 'unfail' so existing recovery handling can re-process them.
 */
class TaskReviewDoubleSpends extends WalletMonitorTask_1.WalletMonitorTask {
    constructor(monitor, triggerMsecs = Monitor_1.Monitor.oneMinute * 12, reviewLimit = 100, minAgeMinutes = 60, triggerQuickMsecs = Monitor_1.Monitor.oneMinute * 1) {
        super(monitor, TaskReviewDoubleSpends.taskName);
        this.triggerMsecs = triggerMsecs;
        this.reviewLimit = reviewLimit;
        this.minAgeMinutes = minAgeMinutes;
        this.triggerQuickMsecs = triggerQuickMsecs;
        this.triggerNextMsecs = this.triggerQuickMsecs;
    }
    trigger(nowMsecsSinceEpoch) {
        return {
            run: TaskReviewDoubleSpends.checkNow ||
                (this.triggerNextMsecs > 0 && nowMsecsSinceEpoch - this.lastRunMsecsSinceEpoch > this.triggerNextMsecs)
        };
    }
    async getLastReviewedCheckpoint() {
        let events = [];
        await this.storage.runAsStorageProvider(async (sp) => {
            events = await sp.findMonitorEvents({
                partial: { event: TaskReviewDoubleSpends.taskName },
                orderDescending: true,
                paged: { limit: 5 }
            });
        });
        for (const event of events) {
            if (!event.details)
                continue;
            try {
                const parsed = JSON.parse(event.details);
                if (typeof parsed.resumeOffset === 'number') {
                    return {
                        resumeOffset: parsed.resumeOffset,
                        expectedProvenTxReqId: parsed.expectedProvenTxReqId
                    };
                }
            }
            catch (_a) {
                continue;
            }
        }
        return undefined;
    }
    async runTask() {
        var _a;
        TaskReviewDoubleSpends.checkNow = false;
        const checkpoint = await this.getLastReviewedCheckpoint();
        const updatedBefore = new Date(Date.now() - this.minAgeMinutes * 60 * 1000);
        const reqs = await this.findReqsToReview(checkpoint, updatedBefore);
        if (reqs.length === 0) {
            this.triggerNextMsecs = this.triggerMsecs;
            return '';
        }
        this.triggerNextMsecs = reqs.length >= this.reviewLimit ? this.triggerQuickMsecs : this.triggerMsecs;
        const reviewed = [];
        const unfails = [];
        let log = ``;
        let lastRetainedDoubleSpendIndex = -1;
        let retainedDoubleSpendCount = 0;
        for (const req of reqs) {
            const gsr = await this.monitor.services.getStatusForTxids([req.txid]);
            const status = (_a = gsr.results[0]) === null || _a === void 0 ? void 0 : _a.status;
            reviewed.push(req);
            if (status !== 'unknown') {
                unfails.push(req.provenTxReqId);
                log += `unfail ${req.provenTxReqId} ${req.txid} status:${status}\n`;
            }
            else {
                lastRetainedDoubleSpendIndex = reviewed.length - 1;
                retainedDoubleSpendCount += 1;
            }
        }
        if (unfails.length > 0) {
            await this.storage.runAsStorageProvider(async (sp) => {
                await sp.updateProvenTxReq(unfails, { status: 'unfail' });
            });
        }
        return JSON.stringify({
            minAgeMinutes: this.minAgeMinutes,
            reviewed: reviewed.length,
            unfails: unfails.length,
            reviewLimit: this.reviewLimit,
            provenTxReqIds: unfails,
            resumeOffset: lastRetainedDoubleSpendIndex >= 0 ? reqs.sourceOffset + retainedDoubleSpendCount - 1 : undefined,
            expectedProvenTxReqId: lastRetainedDoubleSpendIndex >= 0 ? reviewed[lastRetainedDoubleSpendIndex].provenTxReqId : undefined,
            reviewLog: `${reviewed.length} reqs with status 'doubleSpend'\n${log}`
        });
    }
    async findReqsToReview(checkpoint, updatedBefore) {
        var _a;
        let offset = (checkpoint === null || checkpoint === void 0 ? void 0 : checkpoint.resumeOffset) || 0;
        if (checkpoint && checkpoint.expectedProvenTxReqId !== undefined) {
            const verify = await this.storage.findProvenTxReqs({
                partial: { status: 'doubleSpend' },
                paged: { limit: 1, offset: checkpoint.resumeOffset }
            });
            if (((_a = verify[0]) === null || _a === void 0 ? void 0 : _a.provenTxReqId) !== checkpoint.expectedProvenTxReqId) {
                offset = 0;
            }
            else {
                offset += 1;
            }
        }
        const batch = await this.storage.findProvenTxReqs({
            partial: { status: 'doubleSpend' },
            paged: { limit: this.reviewLimit, offset }
        });
        const reqs = batch.filter(req => req.updated_at <= updatedBefore);
        reqs.sourceOffset = offset;
        return reqs;
    }
}
exports.TaskReviewDoubleSpends = TaskReviewDoubleSpends;
TaskReviewDoubleSpends.taskName = 'ReviewDoubleSpends';
TaskReviewDoubleSpends.checkNow = false;
