// @bsv/wallet-toolbox/out/src/services/chaintracker/chaintracks/configureChaintracksIngestors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildChaintracksOptionsWithIngestors = buildChaintracksOptionsWithIngestors;
const BulkIngestorCDNBabbage_1 = require("./Ingest/BulkIngestorCDNBabbage");
const LiveIngestorWhatsOnChainPoll_1 = require("./Ingest/LiveIngestorWhatsOnChainPoll");
const BulkIngestorWhatsOnChainCdn_1 = require("./Ingest/BulkIngestorWhatsOnChainCdn");
/**
 * Builds the shared portion of ChaintracksOptions that all storage backends
 * (Knex, Idb, NoDb) have in common: the options shell and bulk/live ingestors.
 *
 * The caller is responsible for providing the storage implementation.
 */
function buildChaintracksOptionsWithIngestors(params, storage) {
    const { chain, whatsonchainApiKey, maxPerFile, fetch, cdnUrl, addLiveRecursionLimit } = params;
    const co = {
        chain,
        storage,
        bulkIngestors: [],
        liveIngestors: [],
        addLiveRecursionLimit,
        logging: (...args) => console.log(new Date().toISOString(), ...args),
        readonly: false
    };
    const jsonResource = `${chain}NetBlockHeaders.json`;
    const bulkCdnOptions = {
        chain,
        jsonResource,
        fetch,
        cdnUrl,
        maxPerFile
    };
    co.bulkIngestors.push(new BulkIngestorCDNBabbage_1.BulkIngestorCDNBabbage(bulkCdnOptions));
    const wocOptions = {
        chain,
        apiKey: whatsonchainApiKey,
        timeout: 30000,
        userAgent: 'BabbageWhatsOnChainServices',
        enableCache: true,
        chainInfoMsecs: 5000
    };
    const bulkOptions = {
        ...wocOptions,
        jsonResource,
        idleWait: 5000
    };
    co.bulkIngestors.push(new BulkIngestorWhatsOnChainCdn_1.BulkIngestorWhatsOnChainCdn(bulkOptions));
    const liveOptions = {
        ...wocOptions,
        idleWait: 100000
    };
    co.liveIngestors.push(new LiveIngestorWhatsOnChainPoll_1.LiveIngestorWhatsOnChainPoll(liveOptions));
    return co;
}
