// @bsv/wallet-toolbox/out/src/services/chaintracker/chaintracks/createDefaultKnexChaintracksOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDefaultKnexChaintracksOptions = createDefaultKnexChaintracksOptions;
const knex_1 = require("knex");
const ChaintracksFs_1 = require("./util/ChaintracksFs");
const ChaintracksStorageKnex_1 = require("./Storage/ChaintracksStorageKnex");
const ChaintracksFetch_1 = require("./util/ChaintracksFetch");
const BulkFileDataManager_1 = require("./util/BulkFileDataManager");
const configureChaintracksIngestors_1 = require("./configureChaintracksIngestors");
/**
 *
 * @param chain
 * @param rootFolder defaults to "./data/"
 * @returns
 */
function createDefaultKnexChaintracksOptions(chain, rootFolder = './data/', knexConfig, whatsonchainApiKey = '', maxPerFile = 100000, maxRetained = 2, fetch, cdnUrl = 'https://cdn.projectbabbage.com/blockheaders/', liveHeightThreshold = 2000, reorgHeightThreshold = 400, bulkMigrationChunkSize = 500, batchInsertLimit = 400, addLiveRecursionLimit = 36) {
    fetch || (fetch = new ChaintracksFetch_1.ChaintracksFetch());
    const bfo = {
        chain,
        fetch,
        maxPerFile,
        maxRetained,
        fromKnownSourceUrl: cdnUrl
    };
    const bulkFileDataManager = new BulkFileDataManager_1.BulkFileDataManager(bfo);
    if (!knexConfig) {
        knexConfig = {
            client: 'better-sqlite3',
            connection: { filename: ChaintracksFs_1.ChaintracksFs.pathJoin(rootFolder, `${chain}Net_chaintracks.sqlite`) },
            useNullAsDefault: true
        };
    }
    const knexInstance = (0, knex_1.knex)(knexConfig);
    const so = {
        chain,
        knex: knexInstance,
        bulkFileDataManager,
        liveHeightThreshold,
        reorgHeightThreshold,
        bulkMigrationChunkSize,
        batchInsertLimit
    };
    const storage = new ChaintracksStorageKnex_1.ChaintracksStorageKnex(so);
    return (0, configureChaintracksIngestors_1.buildChaintracksOptionsWithIngestors)({ chain, whatsonchainApiKey, maxPerFile, fetch, cdnUrl, addLiveRecursionLimit }, storage);
}
