// @bsv/wallet-toolbox/out/src/services/__tests/getFiatExchangeRate.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv = __importStar(require("dotenv"));
const index_client_1 = require("../../index.client");
const WERR_errors_1 = require("../../sdk/WERR_errors");
dotenv.config();
describe('getFiatExchangeRate service tests', () => {
    jest.setTimeout(99999999);
    function makeFetchedRates(rates) {
        return {
            timestamp: new Date('2026-03-25T00:00:00.000Z'),
            base: 'USD',
            rates
        };
    }
    test('0 uses the chaintracks fiat service without an exchangeratesapi key', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        options.fiatExchangeRates = {
            timestamp: new Date('2020-01-01T00:00:00.000Z'),
            base: 'USD',
            rates: { USD: 1 },
            rateTimestamps: { USD: new Date('2020-01-01T00:00:00.000Z') }
        };
        const services = new index_client_1.Services(options);
        expect(services.updateFiatExchangeRateServices.count).toBe(1);
        expect(services.updateFiatExchangeRateServices.name).toBe('ChaintracksFiatRates');
        const chaintracksService = jest.fn(async () => makeFetchedRates({ EUR: 0.8, GBP: 0.5 }));
        services.updateFiatExchangeRateServices.services[0].service = chaintracksService;
        await expect(services.getFiatExchangeRate('EUR', 'GBP')).resolves.toBeCloseTo(1.6);
        expect(chaintracksService).toHaveBeenCalledWith(['EUR', 'GBP'], services.options);
    });
    test('1 uses only exchangeratesapi when an api key is configured', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        options.exchangeratesapiKey = 'test-api-key';
        options.fiatExchangeRates = {
            timestamp: new Date('2020-01-01T00:00:00.000Z'),
            base: 'USD',
            rates: { USD: 1 },
            rateTimestamps: { USD: new Date('2020-01-01T00:00:00.000Z') }
        };
        const services = new index_client_1.Services(options);
        expect(services.updateFiatExchangeRateServices.count).toBe(1);
        expect(services.updateFiatExchangeRateServices.name).toBe('exchangeratesapi');
        expect(services.updateFiatExchangeRateServices.services.find(s => s.name === 'ChaintracksFiatRates')).toBeUndefined();
        const exchangeratesapiService = jest.fn(async () => makeFetchedRates({ EUR: 0.92 }));
        services.updateFiatExchangeRateServices.services[0].service = exchangeratesapiService;
        await expect(services.getFiatExchangeRate('EUR')).resolves.toBeCloseTo(0.92);
        expect(exchangeratesapiService).toHaveBeenCalledWith(['EUR'], services.options);
    });
    test('2 returns 1 without refreshing when the requested currency matches the base', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        const services = new index_client_1.Services(options);
        const updateFiatExchangeRates = jest.spyOn(services, 'updateFiatExchangeRates');
        await expect(services.getFiatExchangeRate('USD')).resolves.toBe(1);
        expect(updateFiatExchangeRates).not.toHaveBeenCalled();
    });
    test('3 throws when the requested currency rate is unavailable after refresh', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        const services = new index_client_1.Services(options);
        jest
            .spyOn(services, 'updateFiatExchangeRates')
            .mockImplementation(async (_targetCurrencies) => {
            services.options.fiatExchangeRates = {
                ...services.options.fiatExchangeRates,
                rates: {
                    USD: 1
                }
            };
            return services.options.fiatExchangeRates;
        });
        await expect(services.getFiatExchangeRate('EUR')).rejects.toThrow(WERR_errors_1.WERR_INVALID_PARAMETER);
        await expect(services.getFiatExchangeRate('EUR')).rejects.toThrow("valid fiat currency 'EUR' with an exchange rate.");
    });
    test('4 throws when the requested base rate is unavailable after refresh', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        const services = new index_client_1.Services(options);
        jest
            .spyOn(services, 'updateFiatExchangeRates')
            .mockImplementation(async (_targetCurrencies) => {
            services.options.fiatExchangeRates = {
                ...services.options.fiatExchangeRates,
                rates: {
                    EUR: 0.8
                }
            };
            return services.options.fiatExchangeRates;
        });
        await expect(services.getFiatExchangeRate('EUR', 'GBP')).rejects.toThrow(WERR_errors_1.WERR_INVALID_PARAMETER);
        await expect(services.getFiatExchangeRate('EUR', 'GBP')).rejects.toThrow("valid fiat currency 'GBP' with an exchange rate.");
    });
    test('5 chaintracks works against the real service without an exchangeratesapi key', async () => {
        const options = index_client_1.Services.createDefaultOptions('main');
        options.fiatExchangeRates = {
            timestamp: new Date('2020-01-01T00:00:00.000Z'),
            base: 'USD',
            rates: { USD: 1 },
            rateTimestamps: { USD: new Date('2020-01-01T00:00:00.000Z') }
        };
        const services = new index_client_1.Services(options);
        const eurPerUsd = await services.getFiatExchangeRate('EUR');
        expect(eurPerUsd).toBeGreaterThan(0);
        expect(services.updateFiatExchangeRateServices.name).toBe('ChaintracksFiatRates');
        expect(services.options.fiatExchangeRates.rates.EUR).toBeGreaterThan(0);
    });
    test('6 exchangeratesapi works when EXCHANGERATESAPI_KEY is set', async () => {
        const apiKey = process.env.EXCHANGERATESAPI_KEY;
        if (!apiKey)
            return;
        const options = index_client_1.Services.createDefaultOptions('main');
        options.exchangeratesapiKey = apiKey;
        options.fiatExchangeRates = {
            timestamp: new Date('2020-01-01T00:00:00.000Z'),
            base: 'USD',
            rates: { USD: 1 },
            rateTimestamps: { USD: new Date('2020-01-01T00:00:00.000Z') }
        };
        const services = new index_client_1.Services(options);
        const eurPerUsd = await services.getFiatExchangeRate('EUR');
        expect(eurPerUsd).toBeGreaterThan(0);
        expect(services.updateFiatExchangeRateServices.name).toBe('exchangeratesapi');
    });
});
